File: src/lib/money.ts
// Money as integer minor units (cents), not floating point — the
// frontend counterpart to src-tauri/src/money.rs. Every "money"-typed
// field, everywhere in the app, is an integer number of cents on the
// wire and in memory. This file is the ONLY place a decimal string
// gets parsed into cents or cents get formatted back to a decimal
// string — no component does its own `.toFixed(2)` or `parseFloat`
// on a money value anymore.

// Minor-unit decimal places per ISO 4217 currency code. Mirrors
// money::decimal_places_for in the Rust backend exactly — if this
// list ever needs to change, change it in both places.
const ZERO_DECIMAL = new Set([
  'JPY', 'KRW', 'VND', 'UGX', 'RWF', 'XOF', 'XAF', 'BIF', 'DJF', 'GNF',
  'KMF', 'MGA', 'PYG', 'VUV', 'CLP',
]);
const THREE_DECIMAL = new Set(['BHD', 'IQD', 'JOD', 'KWD', 'OMR', 'TND']);

export function decimalPlacesFor(currencyCode: string): number {
  const code = (currencyCode || 'USD').toUpperCase();
  if (ZERO_DECIMAL.has(code)) return 0;
  if (THREE_DECIMAL.has(code)) return 3;
  return 2;
}

/**
 * Formats integer minor units (cents) into a display string, e.g.
 * formatMoney(1250, 'USD') -> "12.50". Currency-aware: a 0-decimal
 * currency like JPY never gets a fractional part appended.
 */
export function formatMoney(cents: number | null | undefined, currencyCode: string = 'USD'): string {
  if (cents === null || cents === undefined || Number.isNaN(cents)) return '';
  const places = decimalPlacesFor(currencyCode);
  const scale = Math.pow(10, places);
  const value = cents / scale;
  return value.toLocaleString(undefined, {
    minimumFractionDigits: places,
    maximumFractionDigits: places,
  });
}

/**
 * Parses a human-typed decimal string into integer minor units.
 * Mirrors money::parse_money_input in the Rust backend: rejects more
 * precision than the currency supports (e.g. "12.505" for a 2dp
 * currency) rather than silently rounding it away. Returns null for
 * anything invalid — callers decide how to surface that.
 *
 * THE BUG THIS FIXES: formatMoney() above renders amounts >= 1000
 * with thousands separators via toLocaleString ("50,000.00") — which
 * is exactly what startEdit() in ModuleView.tsx seeds a money field
 * with when editing an existing record, and exactly what this same
 * function's own onBlur handler re-formats a field into as soon as a
 * person tabs away from it. Every one of those comma-formatted
 * strings then had nowhere to go: this parser rejected any non-digit
 * character, including the very commas formatMoney had just inserted,
 * so simply opening an HR salary (almost always >= 1,000) for edit
 * and saving it straight back — without changing a thing — failed
 * with "'50,000.00' is not a valid amount". Thousands separators are
 * also just how people normally type larger figures by hand. Commas
 * are stripped before every other check runs, so "50,000.00",
 * "50,000", and "50000" all parse identically — this never loosens
 * what counts as a valid decimal point or digit, it only ignores the
 * grouping punctuation a human (or this same file) would put in.
 */
export function parseMoneyInput(input: string, currencyCode: string = 'USD'): number | null {
  const trimmed = (input ?? '').trim().replace(/,/g, '');
  if (!trimmed) return null;

  const negative = trimmed.startsWith('-');
  const unsigned = negative ? trimmed.slice(1) : trimmed;

  const places = decimalPlacesFor(currencyCode);
  const parts = unsigned.split('.');
  if (parts.length > 2) return null;
  const [whole, frac = ''] = parts;

  if (frac.length > places) return null; // more precision than this currency supports
  if (!/^\d*$/.test(whole) || !/^\d*$/.test(frac)) return null;
  if (whole === '' && frac === '') return null;

  const scale = Math.pow(10, places);
  const wholeVal = whole === '' ? 0 : parseInt(whole, 10);
  const fracVal = places === 0 ? 0 : parseInt(frac.padEnd(places, '0'), 10);

  const cents = wholeVal * scale + fracVal;
  if (!Number.isSafeInteger(cents)) return null;
  return negative ? -cents : cents;
}

/**
 * Multiplies an integer cents amount by a quantity — exact, since
 * both are integers. Prefer this over inline `price * qty` so every
 * money calculation in the app is visibly going through one place.
 */
export function multiplyMoney(cents: number, quantity: number): number {
  return cents * quantity;
}

/**
 * Sums an array of integer cents amounts — exact, since JS numbers
 * represent integers up to 2^53 without any floating-point error.
 */
export function sumMoney(amounts: number[]): number {
  return amounts.reduce((total, c) => total + c, 0);
}

File: src-tauri/src/money.rs
//! Money as integer minor units (cents), not floating point.
//!
//! Why this file exists: `f64` cannot exactly represent most decimal
//! currency values, and every multiplication or repeated addition over
//! a money value is a chance for a fraction-of-a-cent error to creep
//! in and compound across thousands of transactions. Storing and
//! computing money as `i64` minor units makes every arithmetic
//! operation exact by construction — there is no rounding error to
//! accumulate because there is never a fractional cent in memory in
//! the first place.
//!
//! This is the ONLY place rounding is allowed to happen. Every other
//! module either passes cents straight through (no rounding needed,
//! because integer math is exact) or calls into here at the specific
//! boundary where a fraction is unavoidable (parsing human input,
//! applying an exchange rate, applying a tax rate).
//!
//! Not every currency has 2 decimal places (JPY has 0, KWD has 3) —
//! `decimal_places_for` is the one place that fact lives, so nothing
//! else hardcodes "divide by 100".

use anyhow::{anyhow, Result};

/// Minor-unit decimal places per ISO 4217 currency code. Defaults to 2
/// (the overwhelmingly common case) for anything not listed here —
/// this list only needs to cover the exceptions.
pub fn decimal_places_for(currency_code: &str) -> u32 {
    match currency_code.to_uppercase().as_str() {
        // Zero-decimal currencies — no minor unit in practice.
        "JPY" | "KRW" | "VND" | "UGX" | "RWF" | "XOF" | "XAF" | "BIF" | "DJF" | "GNF"
        | "KMF" | "MGA" | "PYG" | "VUV" | "CLP" => 0,
        // Three-decimal currencies.
        "BHD" | "IQD" | "JOD" | "KWD" | "OMR" | "TND" => 3,
        // Everything else, including KES and USD: standard 2dp.
        _ => 2,
    }
}

/// Parses a human-typed decimal string (e.g. "12.50", "1250" for a
/// 0dp currency) into integer minor units. Rejects anything with more
/// fractional precision than the currency supports — a sub-cent entry
/// like "12.505" for a 2dp currency is a data-entry mistake, not
/// something to silently round away.
pub fn parse_money_input(input: &str, currency_code: &str) -> Result<i64> {
    // Strip thousands separators before anything else — mirrors
    // src/lib/money.ts's parseMoneyInput exactly (see that file's own
    // doc comment for THE BUG THIS FIXES: formatMoney's own comma
    // grouping used to round-trip straight into a rejection here).
    // A spreadsheet cell can carry the same comma-grouped text a
    // human would type, so this needs the identical fix, not just the
    // frontend.
    let no_separators = input.replace(',', "");
    let trimmed = no_separators.trim();
    if trimmed.is_empty() {
        return Err(anyhow!("amount is required"));
    }
    let negative = trimmed.starts_with('-');
    let unsigned = trimmed.trim_start_matches('-');

    let places = decimal_places_for(currency_code);
    let (whole, frac) = match unsigned.split_once('.') {
        Some((w, f)) => (w, f),
        None => (unsigned, ""),
    };
    if frac.len() > places as usize {
        return Err(anyhow!(
            "'{input}' has more precision than {currency_code} supports ({places} decimal place{})",
            if places == 1 { "" } else { "s" }
        ));
    }
    if !whole.chars().all(|c| c.is_ascii_digit()) || (!frac.is_empty() && !frac.chars().all(|c| c.is_ascii_digit())) {
        return Err(anyhow!("'{input}' is not a valid amount"));
    }
    if whole.is_empty() && frac.is_empty() {
        return Err(anyhow!("'{input}' is not a valid amount"));
    }

    let whole_val: i64 = if whole.is_empty() { 0 } else { whole.parse().map_err(|_| anyhow!("'{input}' is not a valid amount"))? };
    let scale = 10_i64.pow(places);
    let frac_padded = format!("{frac:0<width$}", width = places as usize);
    let frac_val: i64 = if places == 0 { 0 } else { frac_padded.parse().map_err(|_| anyhow!("'{input}' is not a valid amount"))? };

    let cents = whole_val
        .checked_mul(scale)
        .and_then(|v| v.checked_add(frac_val))
        .ok_or_else(|| anyhow!("'{input}' is out of range"))?;

    Ok(if negative { -cents } else { cents })
}

/// Formats integer minor units back to a display string, e.g.
/// 1250 -> "12.50" for a 2dp currency, 1250 -> "1250" for a 0dp one.
pub fn format_money(cents: i64, currency_code: &str) -> String {
    let places = decimal_places_for(currency_code);
    if places == 0 {
        return cents.to_string();
    }
    let scale = 10_i64.pow(places);
    let negative = cents < 0;
    let abs = cents.unsigned_abs() as i64;
    let whole = abs / scale;
    let frac = abs % scale;
    format!(
        "{}{}.{:0width$}",
        if negative { "-" } else { "" },
        whole,
        frac,
        width = places as usize
    )
}

/// Applies a fractional rate (e.g. a tax rate like 0.16, or an
/// exchange rate) to an integer cents amount, rounding to the nearest
/// cent exactly once. This is the ONLY place a rate multiplies a
/// money value — every caller passes the result straight through as
/// an integer afterward, never carrying the unrounded fraction.
pub fn apply_rate(cents: i64, rate: f64) -> i64 {
    (cents as f64 * rate).round() as i64
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_plain_decimal() {
        assert_eq!(parse_money_input("12.50", "USD").unwrap(), 1250);
        assert_eq!(parse_money_input("0.01", "USD").unwrap(), 1);
        assert_eq!(parse_money_input("100", "USD").unwrap(), 10000);
    }

    #[test]
    fn rejects_sub_cent_precision() {
        assert!(parse_money_input("12.505", "USD").is_err());
    }

    #[test]
    fn respects_zero_decimal_currencies() {
        assert_eq!(parse_money_input("1500", "JPY").unwrap(), 1500);
        assert!(parse_money_input("15.50", "JPY").is_err());
    }

    #[test]
    fn respects_three_decimal_currencies() {
        assert_eq!(parse_money_input("1.500", "KWD").unwrap(), 1500);
    }

    #[test]
    fn strips_thousands_separators() {
        assert_eq!(parse_money_input("50,000", "KES").unwrap(), 5_000_000);
        assert_eq!(parse_money_input("50,000.00", "KES").unwrap(), 5_000_000);
        assert_eq!(parse_money_input("1,234,567.89", "USD").unwrap(), 123_456_789);
    }

    #[test]
    fn rejects_garbage() {
        assert!(parse_money_input("abc", "USD").is_err());
        assert!(parse_money_input("", "USD").is_err());
        assert!(parse_money_input("12.5.6", "USD").is_err());
    }

    #[test]
    fn format_round_trips() {
        assert_eq!(format_money(1250, "USD"), "12.50");
        assert_eq!(format_money(1, "USD"), "0.01");
        assert_eq!(format_money(1500, "JPY"), "1500");
        assert_eq!(format_money(-1250, "USD"), "-12.50");
    }

    #[test]
    fn exactness_over_many_operations() {
        // The exact case that would drift with f64: repeated addition
        // of a value that isn't exactly representable in binary.
        let unit_price_cents = 1999_i64; // $19.99
        let qty = 7_i64;
        let mut subtotal = 0_i64;
        for _ in 0..10_000 {
            subtotal = 0;
            subtotal += unit_price_cents * qty;
        }
        assert_eq!(subtotal, 13993);
    }

    #[test]
    fn apply_rate_rounds_once() {
        // 16% tax on 1050 cents = 168.0 exactly
        assert_eq!(apply_rate(1050, 0.16), 168);
        // A case that would show float drift if rounded more than once
        assert_eq!(apply_rate(333, 0.1), 33);
    }
}
